File: hs/p3.hs
iSqrt :: Int -> Int
iSqrt = floor . sqrt . fromIntegral

largestProperFactor :: Int -> Int -> Int
largestProperFactor x y =
    let r = iSqrt x
    in
    if y > r then x
        else if mod x y == 0
            then div x y
            else largestProperFactor x (y + 2)

largestPrimeFactor :: Int -> Int
largestPrimeFactor x
    | x == 2        = 2
    | mod x 2 == 0  = largestPrimeFactor (div x 2)
    | otherwise     =
        let y = largestProperFactor x 3
        in
        if y < x
            then largestPrimeFactor y
            else x

main = let num = 600851475143 in putStrLn (show (largestPrimeFactor num))

File: rs/p3.rs
fn largest_proper_factor(x: u128) -> u128 {
    let sqrt = (x as f64).sqrt().floor() as u128;
    for i in (3..sqrt + 1).step_by(2) {
        if x % i == 0 {
            return x / i
        }
    }
    x
}

fn largest_prime_factor(x: u128) -> u128 {
    let mut y = x;
    while y % 2 == 0 {
        if y == 2 { return 2 }
        y /= 2;
    }

    let mut i = y;
    let mut j = largest_proper_factor(i);
    while i > j {
        i = j;
        j = largest_proper_factor(i);
    }

    i
}

fn main() {
    const NUM: u128 = 600851475143;
    println!("{}", largest_prime_factor(NUM));
}
